### ISSUE TITLE
Failed to reprovision device. IoT Keys Service error - permission denied. could not load key pair. DPS client error. aziot 1.2

### ISSUE THREAD

**Issue by dmaxwe22** [Azure/iotedge#4853]
## Expected Behavior
Would expect X.509 keys to be read on IoT Edge computer and process to work normally.  

## Current Behavior
IoT Edge fails to start. Several errors one after the other.  Could be a file permissions with X.509 certificates issue. Please provide permissions guidance.

From Logs:
"aziot-keyd[8391]: ....  - Permission denied (os error 13)"
"aziot-keyd[8391]: ....  - !!! caused by: could not load key pair: AZIOT_KEYS_RC_ERR_EXTERNAL"
"aziot-keyd[8391]: ....  - --> 500 {"content-type": "application/json"}"

What is best way of uploading X.509 certificates and setting ownership? 

## Steps to Reproduce
Provide a detailed set of steps to reproduce the bug.
1. Install Ubuntu 20.04.2 LTS, follow documentation for both "Create and provision an IoT Edge device using X.509 certificates" and "Install or uninstall Azure IoT Edge for Linux" for IoT Edge 1.2.
2. Apply configuration and view logs.

## Context (Environment)
Using ubuntu server **20.04.2 LTS** with:
  curl https://packages.microsoft.com/config/ubuntu/18.04/multiarch/prod.list > ./microsoft-prod.list

### Output of `iotedge check`

<details>
<summary>Click here</summary>

```
Configuration checks (aziot-identity-service)
---------------------------------------------
√ keyd configuration is well-formed - OK
√ certd configuration is well-formed - OK
√ tpmd configuration is well-formed - OK
√ identityd configuration is well-formed - OK
√ daemon configurations up-to-date with config.toml - OK
√ identityd config toml file specifies a valid hostname - OK
√ aziot-identity-service package is up-to-date - OK
√ host time is close to reference time - OK
√ production readiness: identity certificates expiry - OK
√ preloaded certificates are valid - OK
√ keyd is running - OK
√ certd is running - OK
√ identityd is running - OK
× read all preloaded certificates from the Certificates Service - Error
    could not load cert with ID "aziot-edged-trust-bundle"

    Caused by:
        parameter "id" has an invalid value
        caused by: not found
√ read all preloaded key pairs from the Keys Service - OK
√ ensure all preloaded certificates match preloaded private keys with the same ID - OK

Connectivity checks (aziot-identity-service)
--------------------------------------------
‼ host can connect to and perform TLS handshake with iothub AMQP port - Warning
    Could not retrieve iothub_hostname from provisioning file.
    Please specify the backing IoT Hub name using --iothub-hostname switch if you have that information.
    Since no hostname is provided, all hub connectivity tests will be skipped.
‼ host can connect to and perform TLS handshake with iothub HTTPS / WebSockets port - Warning
    Could not retrieve iothub_hostname from provisioning file.
    Please specify the backing IoT Hub name using --iothub-hostname switch if you have that information.
    Since no hostname is provided, all hub connectivity tests will be skipped.
‼ host can connect to and perform TLS handshake with iothub MQTT port - Warning
    Could not retrieve iothub_hostname from provisioning file.
    Please specify the backing IoT Hub name using --iothub-hostname switch if you have that information.
    Since no hostname is provided, all hub connectivity tests will be skipped.
√ host can connect to and perform TLS handshake with DPS endpoint - OK

Configuration checks
--------------------
√ aziot-edged configuration is well-formed - OK
√ configuration up-to-date with config.toml - OK
√ container engine is installed and functional - OK

```
</details>

### Device Information
* Host OS [e.g. Ubuntu 18.04, Windows Server IoT 2019]: 
* Architecture [e.g. amd64, arm32, arm64]:  x86_64
* Container OS [e.g. Linux containers, Windows containers]:  Linux containers

### Runtime Versions
* aziot-edged [run `iotedge version`]: 1.2.0 
* Edge Agent [image tag (e.g. 1.0.0)]:  N/A 
* Edge Hub [image tag (e.g. 1.0.0)]: N/A
* Docker/Moby [run `docker version`]:  20.10.6+azure

## Logs

<details>
<summary>DPS section in config.toml</summary>

```

# DPS provisioning with X.509 certificate
[provisioning]
source = "dps"
global_endpoint = "https://global.azure-devices-provisioning.net"
id_scope = "0nexxxxxxxxxx" (hidden)

[provisioning.attestation]
method = "x509"
registration_id = "IoT-Edge-knuser-ID"

identity_cert = "file:///etc/aziot/certs/identity.pem"                # file URI, or...

identity_pk = "file:///etc/aziot/certs/identity.key"                    # file URI, or...


```
</details>

<details>
<summary>iotedge system logs</summary>

```

tsiteb@ubuntu-knuser:/etc/aziot$ sudo iotedge system logs
-- Logs begin at Thu 2021-04-15 02:26:31 UTC, end at Thu 2021-04-15 17:01:59 UTC. --
Apr 15 17:01:46 ubuntu-knuser systemd[1]: Started Azure IoT Edge daemon.
Apr 15 17:01:46 ubuntu-knuser aziot-edged[8332]: 2021-04-15T17:01:46Z [INFO] - Starting Azure IoT Edge Module Runtime
Apr 15 17:01:46 ubuntu-knuser aziot-edged[8332]: 2021-04-15T17:01:46Z [INFO] - Version - 1.2.0
Apr 15 17:01:46 ubuntu-knuser aziot-edged[8332]: 2021-04-15T17:01:46Z [INFO] - Initializing the module runtime...
Apr 15 17:01:46 ubuntu-knuser aziot-edged[8332]: 2021-04-15T17:01:46Z [INFO] - Initializing module runtime...
Apr 15 17:01:46 ubuntu-knuser aziot-edged[8332]: 2021-04-15T17:01:46Z [INFO] - Using runtime network id azure-iot-edge
Apr 15 17:01:46 ubuntu-knuser aziot-edged[8332]: 2021-04-15T17:01:46Z [INFO] - Successfully initialized module runtime
Apr 15 17:01:46 ubuntu-knuser aziot-edged[8332]: 2021-04-15T17:01:46Z [INFO] - Finished initializing the module runtime.
Apr 15 17:01:46 ubuntu-knuser aziot-edged[8332]: 2021-04-15T17:01:46Z [INFO] - Obtaining edge device provisioning data...
Apr 15 17:01:46 ubuntu-knuser systemd[1]: Started Azure IoT Identity Service.
Apr 15 17:01:46 ubuntu-knuser aziot-identityd[8378]: 2021-04-15T17:01:46Z [INFO] - Starting service...
Apr 15 17:01:46 ubuntu-knuser aziot-identityd[8378]: 2021-04-15T17:01:46Z [INFO] - Version - 1.2.0
Apr 15 17:01:46 ubuntu-knuser aziot-identityd[8378]: 2021-04-15T17:01:46Z [INFO] - Provisioning starting. Reason: Startup
Apr 15 17:01:46 ubuntu-knuser systemd[1]: Started Azure IoT Certificates Service.
Apr 15 17:01:46 ubuntu-knuser aziot-certd[8383]: 2021-04-15T17:01:46Z [INFO] - Starting service...
Apr 15 17:01:46 ubuntu-knuser aziot-certd[8383]: 2021-04-15T17:01:46Z [INFO] - Version - 1.2.0
Apr 15 17:01:46 ubuntu-knuser aziot-certd[8383]: 2021-04-15T17:01:46Z [INFO] - Starting server...
Apr 15 17:01:46 ubuntu-knuser aziot-certd[8383]: 2021-04-15T17:01:46Z [INFO] - <-- GET /certificates/device-id?api-version=2020-09-01 {"host": "certd.sock"}
Apr 15 17:01:46 ubuntu-knuser aziot-certd[8383]: 2021-04-15T17:01:46Z [INFO] - --> 200 {"content-type": "application/json"}
Apr 15 17:01:46 ubuntu-knuser systemd[1]: Started Azure IoT Keys Service.
Apr 15 17:01:46 ubuntu-knuser aziot-keyd[8391]: 2021-04-15T17:01:46Z [INFO] - Starting service...
Apr 15 17:01:46 ubuntu-knuser aziot-keyd[8391]: 2021-04-15T17:01:46Z [INFO] - Version - 1.2.0
Apr 15 17:01:46 ubuntu-knuser aziot-keyd[8391]: 2021-04-15T17:01:46Z [INFO] - Loaded libaziot-keys with version 0x02000000
Apr 15 17:01:46 ubuntu-knuser aziot-keyd[8391]: 2021-04-15T17:01:46Z [INFO] - Starting server...
Apr 15 17:01:46 ubuntu-knuser aziot-keyd[8391]: 2021-04-15T17:01:46Z [INFO] - <-- GET /keypair/device-id?api-version=2020-09-01 {"host": "keyd.sock"}
Apr 15 17:01:46 ubuntu-knuser aziot-keyd[8391]: 2021-04-15T17:01:46Z [ERR!] - Permission denied (os error 13)
Apr 15 17:01:46 ubuntu-knuser aziot-keyd[8391]: 2021-04-15T17:01:46Z [ERR!] - !!! internal error
Apr 15 17:01:46 ubuntu-knuser aziot-keyd[8391]: 2021-04-15T17:01:46Z [ERR!] - !!! caused by: could not load key pair
Apr 15 17:01:46 ubuntu-knuser aziot-keyd[8391]: 2021-04-15T17:01:46Z [ERR!] - !!! caused by: could not load key pair: AZIOT_KEYS_RC_ERR_EXTERNAL
Apr 15 17:01:46 ubuntu-knuser aziot-keyd[8391]: 2021-04-15T17:01:46Z [INFO] - --> 500 {"content-type": "application/json"}
Apr 15 17:01:46 ubuntu-knuser aziot-identityd[8378]: 2021-04-15T17:01:46Z [WARN] - Failed to reprovision device. Running offline. Reprovisioning failure reason: DPS client error.
Apr 15 17:01:46 ubuntu-knuser aziot-identityd[8378]: 2021-04-15T17:01:46Z [INFO] - Starting server...
Apr 15 17:01:46 ubuntu-knuser aziot-identityd[8378]: 2021-04-15T17:01:46Z [INFO] - <-- POST /identities/device?api-version=2020-09-01 {"content-type": "application/json", "host": "2f72756e2f617a696f742f6964656e74697479642e736f636b:0", "content-length": "16"}
Apr 15 17:01:46 ubuntu-knuser aziot-identityd[8378]: 2021-04-15T17:01:46Z [INFO] - !!! device identity not found
Apr 15 17:01:46 ubuntu-knuser aziot-identityd[8378]: 2021-04-15T17:01:46Z [INFO] - --> 400 {"content-type": "application/json"}
Apr 15 17:01:46 ubuntu-knuser aziot-edged[8332]: 2021-04-15T17:01:46Z [WARN] - The daemon could not start up successfully: Could not retrieve device information
Apr 15 17:01:46 ubuntu-knuser aziot-edged[8332]: 2021-04-15T17:01:46Z [WARN] -         caused by: HTTP response error: [400 Bad Request] {"message":"device identity not found"}
Apr 15 17:01:51 ubuntu-knuser aziot-edged[8332]: 2021-04-15T17:01:51Z [WARN] - Retrying getting edge device provisioning information.
Apr 15 17:01:51 ubuntu-knuser aziot-edged[8332]: 2021-04-15T17:01:51Z [INFO] - Obtaining edge device provisioning data...
Apr 15 17:01:51 ubuntu-knuser aziot-edged[8332]: 2021-04-15T17:01:51Z [WARN] - The daemon could not start up successfully: Could not retrieve device information
Apr 15 17:01:51 ubuntu-knuser aziot-edged[8332]: 2021-04-15T17:01:51Z [WARN] -         caused by: HTTP response error: [400 Bad Request] {"message":"device identity not found"}


```
</details>

<details>
<summary>iotedge system status</summary>

```
tsiteb@ubuntu-knuser:/etc/aziot$ sudo iotedge system status
System services:
    aziot-edged             Running
    aziot-identityd         Running
    aziot-keyd              Running
    aziot-certd             Running
    aziot-tpmd              Ready

```
</details>

**Comment by lfitchett** [comment]
```× read all preloaded certificates from the Certificates Service - Error
    could not load cert with ID "aziot-edged-trust-bundle"

    Caused by:
        parameter "id" has an invalid value
        caused by: not found```
        
This means that the preloaded cert's file path is wrong. Make sure the preloaded certificate section of the config.toml points to a valid file path. If the certificate is for example located in /foo/bar/baz.pem, the value should be 
aziot-edged-trust-bundle="file:///foo/bar/baz.pem"

**Comment by dmaxwe22** [comment]
The problem turned out to be an ownership or permissions problem related to the private key. 

Since the creation of the certificate was done through a web call to a trusted global CA, the certificate permissions were OK but the private key permissions were created using openssl and the sudo root user. So, the ownership of the private key was root.  Looking at the logs I could see that the certificate had good enough permissions and so I copied the permissions over to the key.  Actually the first thing I did was to rename the key just to be safe. I renamed identity.key to identity-key.pem. Then did the following:

sudo chmod --reference=identity.pem identity-key.pem

This changed ownership of both certificate and key to show the following: 
-rw-rw-r--

On a different subject, just FYI for others as documentation has not quite caught up here...  there was still another problem. 

The X.509 identity certificate CN field needs to match the config.toml registration_id field. So, in my case CN=IoT-Edge-knuser-SuperLogics-ID and so registration_id had to be changed to the same 
registration_id="IoT-Edge-knuser-SuperLogics-ID"

Now there are no more errors, the device has been provisioned and enrolled and data is streaming into the Azure IoT Hub.
